### PR TITLE
Add Indexer and Esplora API layers

### LINKED ISSUE #25: Indexer & Esplora API layer
### Describe the feature

**Description:**
Create typed API clients for Indexer and Esplora with React Query hooks and error handling.

**Acceptance Criteria:**
- [x] Indexer API client ported from `web/src/api/client.ts` with Zod schemas for:
  - [x] GET `/offers` (with status filter)
  - [x] GET `/offers/:id` (full details)
  - [x] GET `/offers/:id/utxos`
  - [x] GET `/offers/:id/participants`
  - [x] POST `/offers/batch`
  - [x] GET `/offers/by-script`
  - [x] GET `/offers/by-borrower-pubkey`
- [x] Esplora API client ported from `web/src/api/esplora.ts` with methods:
  - [x] `getTx(txid)` — fetch transaction
  - [x] `broadcastTx(txHex)` — broadcast raw transaction
  - [x] `getAddressInfo(addr)` — tx count, balance
  - [x] `getAddressUtxo(addr)` — UTXO list
  - [x] `getLatestBlockHeight()` — current height
  - [x] `getBlockHashAtHeight(height)`
- [x] React Query hooks created:
  - [x] `useOffers(params)` — list offers
  - [x] `useOffer(id)` — single offer
  - [x] `useOfferUtxos(id)` — offer UTXOs
  - [x] `useOfferParticipants(id)` — participants
  - [x] `useAddressUtxos(address)` — address UTXOs
  - [x] `useBlockHeight()` — current block height
  - [x] `useAddressTxs(address)` — address tx history
- [x] Centralized error handling with Zod validation
- [x] Exponential backoff retry strategy (max 3 retries)
- [x] React Query DevTools shows typed responses
- [x] Zod validation catches malformed API responses and logs errors
- [x] API base URLs from `VITE_API_URL` and `VITE_ESPLORA_BASE_URL` env vars

**Dependencies:**
- Requires #22 complete

**Notes:**
- BigInt normalization for u64 amounts/heights from API
- Explorer URL generation for Esplora links
- All queries configured with stale time and cache invalidation

### PR DESCRIPTION
**Pull request by nikita1tup**
Closes #25

### CODE REVIEW

File: web-v2/package.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
   "dependencies": {
     "@heroui/react": "^3.0.4",
     "@tanstack/react-query": "^5.100.10",
+    "@tanstack/react-query-devtools": "^5.100.10",
```
Comment (lilbonekit) [inline web-v2/package.json]: Should be moved to `devDependencies`.

File: web-v2/src/constants/env.ts [hunk 2]
```diff
@@ -3,13 +3,18 @@ import { z } from 'zod'
...
   VITE_API_URL: import.meta.env.VITE_API_URL,
   VITE_ESPLORA_BASE_URL: import.meta.env.VITE_ESPLORA_BASE_URL,
+  VITE_ESPLORA_EXPLORER_URL: import.meta.env.VITE_ESPLORA_EXPLORER_URL,
   VITE_NETWORK: import.meta.env.VITE_NETWORK,
 })
 
 export type AppEnv = z.infer<typeof envSchema>
+
+export const IS_DEV: boolean = import.meta.env.DEV
+export const IS_PROD: boolean = import.meta.env.PROD
```
Comment (lilbonekit) [inline web-v2/src/constants/env.ts]: `IS_DEV` and `IS_PROD` bypass the centralized env layer. Feels inconsistent with the rest of the env access pattern.

File: web-v2/src/api/schemas.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+function coerceToNumber(value: unknown): unknown {
+  if (value === null || value === undefined) return 0
+  if (typeof value === 'number') return value
+  if (typeof value === 'string') return Number(value)
+  return value
+}
+
+const u64AsBigint = z.preprocess(coerceToBigint, z.bigint())
+const u64AsNumber = z.preprocess(coerceToNumber, z.number())
```
Comment (lilbonekit) [inline web-v2/src/api/schemas.ts]: `u64AsNumber` is a bit misleading since JS numbers cannot safely represent arbitrary u64 values.

File: web-v2/src/api/schemas.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  'repaid',
+  'liquidated',
+  'cancelled',
+  'claimed',
+])
+export type OfferStatus = z.infer<typeof offerStatusSchema>
+
+const offerStatusWithFallback = offerStatusSchema.catch(ctx => {
+  console.warn('[api] unknown offer status, falling back to "pending"', { value: ctx.input })
+  return 'pending'
```
Comment (lilbonekit) [inline web-v2/src/api/schemas.ts]: This fallback turns unknown backend states into valid UI states

File: web-v2/src/api/schemas.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    asset: z.string().optional(),
+  })
+  .passthrough()
+export type EsploraVout = z.infer<typeof esploraVoutSchema>
+
+export const esploraTxSchema = z
+  .object({
+    txid: z.string(),
+    vout: z.array(esploraVoutSchema),
+    vin: z.array(z.unknown()).optional(),
```
Comment (lilbonekit) [inline web-v2/src/api/schemas.ts]: Any reason `vin` is intentionally typed as `unknown[]`?

File: web-v2/src/api/queryKeys.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { normalizeHex } from '@/utils/hex'
+
+import type { ListOffersParams } from './indexer'
+
+export const queryKeys = {
+  offers: {
+    all: ['offers'] as const,
+    list: (params: ListOffersParams = {}) => ['offers', 'list', params] as const,
```
Comment (lilbonekit) [inline web-v2/src/api/queryKeys.ts]: Using objects directly in query keys can sometimes be problematic

File: web-v2/src/api/indexer.ts [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  return requestJson(`${indexerBaseUrl()}/offers/batch`, schema, {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify({ ids }),
+    signal: options.signal,
+  })
+}
+
+export interface ListOffersParams {
+  status?: OfferStatus | string
```
Comment (lilbonekit) [inline web-v2/src/api/indexer.ts]: Why is `status` typed as `OfferStatus | string` instead of just `OfferStatus`?

File: web-v2/src/api/indexer.ts [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+const offerShortListSchema = z.array(offerShortSchema)
+const offerFullListSchema = z.array(offerFullSchema)
+const offerDetailsListSchema = z.array(offerDetailsSchema)
+const offerUtxoListSchema = z.array(offerUtxoSchema)
+const offerParticipantListSchema = z.array(offerParticipantSchema)
+
+function indexerBaseUrl(): string {
+  return env.VITE_API_URL.replace(/\/+$/, '')
+}
```
Comment (lilbonekit) [inline web-v2/src/api/indexer.ts]: This could probably just be a normalized constant instead of a function since the env value is static

File: web-v2/src/api/esplora.ts [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+  scripthashUtxoEntrySchema,
+} from './schemas'
+import { requestBytes, requestJson, type RequestParams, requestText } from './transport'
+
+const txidListSchema = z.array(z.string())
+const utxoListSchema = z.array(scripthashUtxoEntrySchema)
+const txListSchema = z.array(scripthashTxEntrySchema)
+const outspendsListSchema = z.array(esploraOutspendSchema)
+
+function esploraBaseUrl(): string {
```
Comment (lilbonekit) [inline web-v2/src/api/esplora.ts]: Same case as `indexerBaseUrl`

File: web-v2/src/api/indexer/hooks.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  })
+}
+
+export function useOffer(offerId: string): UseQueryResult<OfferDetails> {
+  return useQuery({
+    queryKey: queryKeys.offers.detail(offerId),
+    queryFn: ({ signal }) => fetchOffer(offerId, { signal }),
+    staleTime: STALE_TIME_MS.realtime,
+  })
+}
```
Comment (lilbonekit) [inline web-v2/src/api/indexer/hooks.ts:32]: Should these queries use `enabled` guards for empty/invalid ids?

File: web-v2/src/api/hooks/esplora.ts
Comment (lilbonekit) [inline web-v2/src/api/hooks/esplora.ts:1]: Some of these queries (`useTx`, `useAddressInfo`, `useAddressUtxos`, etc.) probably should use `enabled` guards for empty or invalid params

File: web-v2/src/constants/env.ts [hunk 11]
```diff
@@ -1,15 +1,21 @@
 import { z } from 'zod'
 
 const envSchema = z.object({
-  VITE_API_URL: z.string().url().default('http://localhost:8000'),
+  VITE_API_URL: z.string().url().default('http://localhost:80'),
   VITE_ESPLORA_BASE_URL: z.string().url().default('https://blockstream.info/liquidtestnet/api'),
+  VITE_ESPLORA_EXPLORER_URL: z.string().url().default('https://blockstream.info/liquidtestnet'),
   VITE_NETWORK: z.enum(['liquid', 'liquidtestnet', 'regtest']).default('liquidtestnet'),
```
Comment (ardier16) [inline web-v2/src/constants/env.ts:6]: Update `.env.example` too

File: web-v2/src/constants/env.ts [hunk 12]
```diff
@@ -1,15 +1,21 @@
 import { z } from 'zod'
 
 const envSchema = z.object({
-  VITE_API_URL: z.string().url().default('http://localhost:8000'),
+  VITE_API_URL: z.string().url().default('http://localhost:80'),
   VITE_ESPLORA_BASE_URL: z.string().url().default('https://blockstream.info/liquidtestnet/api'),
+  VITE_ESPLORA_EXPLORER_URL: z.string().url().default('https://blockstream.info/liquidtestnet'),
```
Comment (ardier16) [inline web-v2/src/constants/env.ts]: Can we merge these two? E.g. `VITE_ESPLORA_BASE_URL=https://blockstream.info/liquidtestnet` and use `${VITE_ESPLORA_BASE_URL}/api` for API calls

**Review by lilbonekit — APPROVED** [review]

File: web-v2/src/providers/queryClient.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
+import { QueryClient } from '@tanstack/react-query'
+
+import { computeRetryDelay, shouldRetryQuery } from '@/api/retry'
```
Comment (ardier16) [inline web-v2/src/providers/queryClient.ts]: Let's include these retry utils inline here in query client. They are small enough and QueryClient scoped

File: web-v2/src/api/index.ts
Comment (ardier16) [inline web-v2/src/api/index.ts:1]: No need for index file, let's import modules explicitly e.g.:
`import { useOffers } from '@/api/hooks/offers`
instead of
`import { useOffers } from '@/api`

That way we can easily see where this `useOffers` comes from.

File: web-v2/src/api/retry.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+function isRetryableError(error: unknown): boolean {
+  if (!(error instanceof ApiError) || error instanceof ApiValidationError) return false
+  const { status } = error
+  return status === undefined || status === 408 || status === 429 || status >= 500
+}
+
+export function shouldRetryQuery(failureCount: number, error: unknown): boolean {
+  if (failureCount >= MAX_RETRY_ATTEMPTS) return false
+  return isRetryableError(error)
```
Comment (ardier16) [inline web-v2/src/api/retry.ts]:
```suggestion
  return failureCount < MAX_RETRY_ATTEMPTS && isRetryableError(error)
```

File: web-v2/src/api/schemas.ts [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+export const addressInfoSchema = z
+  .object({
+    address: z.string(),
+    chain_stats: chainOrMempoolStatsSchema,
+    mempool_stats: chainOrMempoolStatsSchema,
+  })
+  .passthrough()
+export type AddressInfo = z.infer<typeof addressInfoSchema>
+
+export const scripthashUtxoEntrySchema = z
```
Comment (ardier16) [inline web-v2/src/api/schemas.ts]: `scripthash` -> `scriptHash` everywhere

File: web-v2/src/api/esplora.ts
Comment (ardier16) [inline web-v2/src/api/esplora.ts:1]: Let's group API folder by modules:
```
api/
    esplora/
        queryKeys.ts / constants.ts
        hooks.ts (or a folder if many hooks)
        schemas.ts
        calls.ts / functions.ts / methods.ts (need to think about naming)

    indexer/
        queryKeys.ts
        hooks.ts
        schemas.ts
        calls.ts

    errors.ts
    constants.ts
    ...other shared files
```

This way we don't mess different modules in a single scope (e.g. esplora and indexer in queryKeys). More than that this approach allows us to add/remove API modules easily providing better scalability.

File: web-v2/src/api/hooks/esplora.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    queryKey: queryKeys.esplora.addressUtxo(address),
+    queryFn: ({ signal }) => getAddressUtxo(address, { signal }),
+    staleTime: STALE_TIME_MS.realtime,
+    enabled: address.length > 0,
+  })
+}
+
+export function useAddressTxs(
+  address: string,
+  lastSeenTxid?: string,
```
Comment (ardier16) [inline web-v2/src/api/hooks/esplora.ts]: TxId

File: web-v2/src/api/hooks/esplora.ts [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+  })
+}
+
+export function useBroadcastTx(): UseMutationResult<string, ApiError, string> {
+  const queryClient = useQueryClient()
+  return useMutation({
+    mutationFn: (txHex: string) => broadcastTx(txHex),
+    onSuccess: () => {
+      void queryClient.invalidateQueries({ queryKey: queryKeys.esplora.all })
+      void queryClient.invalidateQueries({ queryKey: queryKeys.offers.all })
```
Comment (ardier16) [inline web-v2/src/api/hooks/esplora.ts]: Offers query shouldn't be a part of broadcast tx, let's pass `onSuccess` prop instead

File: web-v2/src/api/client.ts
Comment (ardier16) [inline web-v2/src/api/client.ts:1]: Let's try to use `axios` or another common HTTP library, no need to reinvent a transport layer. We can fall into reliability issues later, cause we have no tests and not all cases covered.

If we can stick with a lib that already fixed most of the edge cases we'll avoid nasty debugging of weird HTTP client issues.

File: web-v2/src/api/indexer.ts [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+  offerUtxoSchema,
+  type ParticipantType,
+} from './schemas'
+import { requestJson, type RequestParams } from './transport'
+
+const offerShortListSchema = z.array(offerShortSchema)
+const offerFullListSchema = z.array(offerFullSchema)
+const offerDetailsListSchema = z.array(offerDetailsSchema)
+const offerUtxoListSchema = z.array(offerUtxoSchema)
+const offerParticipantListSchema = z.array(offerParticipantSchema)
```
Comment (ardier16) [inline web-v2/src/api/indexer.ts]: Why not in `schemas` file?

File: web-v2/src/api/indexer.ts [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+    (left, right) => right.created_at_height - left.created_at_height,
+  )
+  return sortedByHeight[0] ?? null
+}
+
+export function getCurrentLenderParticipant(history: OfferParticipant[]): OfferParticipant | null {
+  return getCurrentParticipantByRole(history, 'lender')
+}
+
+export function getCurrentBorrowerParticipant(
```
Comment (ardier16) [inline web-v2/src/api/indexer.ts]: Some functions prefixed with `fetch...`, some with `get...`, let's choose a consistent one

File: web-v2/src/api/indexer.ts [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+  history: OfferParticipant[],
+): OfferParticipant | null {
+  return getCurrentParticipantByRole(history, 'borrower')
```
Comment (ardier16) [inline web-v2/src/api/indexer.ts]: Seems like a useless method, I can call `getCurrentParticipantByRole(history, 'borrower')` directly

File: web-v2/src/api/indexer.ts [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+): OfferDetails[] {
+  const targetScript = normalizeHex(scriptPubkeyHex)
+  return offers.filter(offer =>
+    offer.participants.some(
+      participant =>
+        participant.participant_type === role &&
+        normalizeHex(participant.script_pubkey) === targetScript,
+    ),
+  )
+}
```
Comment (ardier16) [inline web-v2/src/api/indexer.ts]: It can be moved to `/api/indexer/utils.ts` or `/api/indexer/helpers.ts` with new structure approach

File: web-v2/src/api/esplora.ts [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+export async function getScripthashUtxo(
+  scripthash: string,
+  options: RequestParams = {},
+): Promise<ScripthashUtxoEntry[]> {
+  return fetchResourceUtxos('scripthash', scripthash, options)
+}
+
+export async function getScripthashTxs(
```
Comment (ardier16) [inline web-v2/src/api/esplora.ts]: ScriptHash

File: web-v2/src/api/schemas.ts [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+  return value
+}
+
+const u64AsBigint = z.preprocess(coerceToBigint, z.bigint())
+const blockHeightSchema = z.preprocess(
+  coerceToNumber,
+  z.number().int().nonnegative().max(Number.MAX_SAFE_INTEGER),
+)
+
+const finiteNumber = z.coerce.number().refine(Number.isFinite, 'must be finite')
```
Comment (ardier16) [inline web-v2/src/api/schemas.ts]: We can extract these to `/src/utils/zod.ts` if they are widely used

**Review by ardier16 — CHANGES_REQUESTED** [review]

File: web-v2/src/api/esplora/methods.ts [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+  })
+  return data.trim()
+}
+
+export async function fetchLatestBlockHash(options: RequestParams = {}): Promise<string> {
+  const { data } = await apiClient.get<string>(buildEsploraUrl('/blocks/tip/hash'), {
+    responseType: 'text',
+    signal: options.signal,
+  })
+  return data.trim()
```
Comment (ardier16) [inline web-v2/src/api/esplora/methods.ts]: let's have `requestJson` and `requestText` inside transport file, then we don't need to expose `apiClient` and `parseWithSchema`

File: web-v2/src/api/client.ts
Comment (ardier16) [inline web-v2/src/api/client.ts:1]: What do you think about renaming `transport.ts` -> `client.ts`? Does it make sense?

**Review by ardier16 — APPROVED** [review]

**Comment by ardier16** [comment]
@Hrom131 could you review `nginx/nginx.conf` changes please?

**Review by Hrom131 — APPROVED** [review]
ACK
